File: split/printromanint.pas
procedure printromanint(n: integer);
label
  10;
var
  j, k: poolpointer;
  u, v: nonnegativeinteger;
begin
  j := strstart[260];
  v := 1000;
  while true do
  begin
    while n >= v do
    begin
      printchar(strpool[j]);
      n := n - v;
    end;
    if n <= 0 then
      goto 10;
    k := j + 2;
    u := v div (strpool[k - 1] - 48);
    if strpool[k - 1] = 50 then
    begin
      k := k + 2;
      u := u div (strpool[k - 1] - 48);
    end;
    if n + u >= v then
    begin
      printchar(strpool[k]);
      n := n + u;
    end
    else
    begin
      j := j + 2;
      v := v div (strpool[j - 1] - 48);
    end;
  end;
  10: ;
end;

File: cpp/src/printromanint.cpp
#include "printromanint.h"
#include "printchar.h"

void printromanint(int n)
{
	static const std::string s = "m2d5c2l5x2v5i";
	int j = 0;
	int v = 1000;
	while (true)
	{
		while (n >= v)
		{
			printchar(s[j]);
			n -= v;
		}
		if (n <= 0)
			return;
		int k = j+2;
		int u = v/(s[j+1]-'0');
		if (s[j+1] == '2')
		{
			k += 2;
			u /= s[j+3]-'0';
		}
		if (n+u >= v)
		{
			printchar(s[k]);
			n += u;
		}
		else
		{
			j += 2;
			v /= s[j-1]-'0';
		}
	}
}
